core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLRecordReader.java

```diff
@@ -0,0 +1,305 @@
+    actualNodeLevelMap = new HashMap<Integer, String>();
+    for (int i = 0; i < totalNodes; i++) {
+      actualNodeLevelMap.put(i, nodes[i]);
+    }
+    currentNodeLevelMap = new HashMap<Integer, String>();
+
+    tempFilePath = conf.get(XMLInputFormat.XML_INPUTFORMAT_PROCESSED_DATA_TEMP_FOLDER);
+    fileAction = conf.get(XMLInputFormat.XML_INPUTFORMAT_FILE_ACTION);
+    targetFolder = conf.get(XMLInputFormat.XML_INPUTFORMAT_TARGET_FOLDER);
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (reader != null) {
+      try {
+        reader.close();
+      } catch (XMLStreamException exception) {
+        // Swallow exception.
+      }
+    }
+  }
+
+  @Override
+  public float getProgress() throws IOException {
+    float progress = (float) inputStream.getTotalBytes() /  availableBytes;
+    progress = Float.valueOf(df.format(progress));
+    return progress;
+  }
+
+  @Override
+  public LongWritable getCurrentKey() throws IOException, InterruptedException {
+    return currentKey;
+  }
+
+  @Override
+  public Map<String, String> getCurrentValue() throws IOException, InterruptedException {
+    return currentValue;
+  }
+
+  @Override
+  public void initialize(InputSplit split, TaskAttemptContext context) throws IOException, InterruptedException {
+  }
+
+  @Override
+  public boolean nextKeyValue() throws IOException, InterruptedException {
+    currentKey = new LongWritable();
+    currentValue = new HashMap<String, String>();
+    int lastNodeIndex = totalNodes - 1;
+    String lastNode = nodes[lastNodeIndex];
+    StringBuilder xmlRecord = new StringBuilder();
+
+    //flag to know if xml record matching to node path has been read and ready to use.
+    boolean xmlRecordReady = false;
+    //flag to know if matching node found as per node path
+    boolean nodeFound = false;
+
+    try {
+      while (reader.hasNext()) {
+        int event = reader.next();
+        switch (event) {
+          case XMLStreamConstants.START_ELEMENT:
+            String nodeNameStart = reader.getLocalName();
+            boolean validHierarchy = true;
+            currentNodeLevelMap.put(nodeLevel, nodeNameStart);
+
+            //check if node hierarchy matches with expected one.
+            for (int j = nodeLevel; j >= 0; j--) {
+              if (!currentNodeLevelMap.get(j).equals(actualNodeLevelMap.get(j))) {
+                validHierarchy = false;
+              }
+            }
+            //check if node hierarchy is valid and it matches with last node in node path
+            //then start appending tag information to create valid XML Record.
+            if (validHierarchy && nodeNameStart.equals(lastNode)) {
+              appendStartTagInformation(nodeNameStart, xmlRecord);
+              //Set flag for valid node path found
+              nodeFound = true;
+              //set file offset
+              currentKey.set(reader.getLocation().getLineNumber());
+            } else if (nodeFound) {
```

Comment:
`nodeFound` turns true only when validHeirarchy is true and this else if block is to process all child element of last node. 
And `nodeFound` turns false when end tag of last node encounters.